src/zon.zig

```diff
@@ -0,0 +1,661 @@
+                .storage = .{
+                    .bytes = try ip.getOrPutString(gpa, self.sema.pt.tid, bytes.items, .maybe_embedded_nulls),
+                },
+            } });
+            const ptr_ty = try self.sema.pt.ptrTypeSema(.{
+                .child = array_ty.toIntern(),
+                .flags = .{
+                    .alignment = .none,
+                    .is_const = true,
+                    .address_space = .generic,
+                },
+            });
+            return self.sema.pt.intern(.{ .ptr = .{
+                .ty = ptr_ty.toIntern(),
+                .base_addr = .{ .anon_decl = .{
+                    .val = array_val,
+                    .orig_ty = ptr_ty.toIntern(),
+                } },
+                .byte_offset = 0,
+            } });
+        },
+        .multiline_string_literal => {
+            var bytes = std.ArrayListUnmanaged(u8){};
+            defer bytes.deinit(gpa);
+
+            var parser = std.zig.string_literal.multilineParser(bytes.writer(gpa));
+            var tok_i = data[node].lhs;
+            while (tok_i <= data[node].rhs) : (tok_i += 1) {
+                try parser.line(self.file.tree.tokenSlice(tok_i));
+            }
+
+            const array_ty = try self.sema.pt.arrayType(.{ .len = bytes.items.len, .sentinel = .zero_u8, .child = .u8_type });
+            const array_val = try self.sema.pt.intern(.{ .aggregate = .{
+                .ty = array_ty.toIntern(),
+                .storage = .{
+                    .bytes = (try ip.getOrPutString(gpa, self.sema.pt.tid, bytes.items, .no_embedded_nulls)).toString(),
+                },
+            } });
+            const ptr_ty = try self.sema.pt.ptrTypeSema(.{
+                .child = array_ty.toIntern(),
+                .flags = .{
+                    .alignment = .none,
+                    .is_const = true,
+                    .address_space = .generic,
+                },
+            });
+            return self.sema.pt.intern(.{ .ptr = .{
+                .ty = ptr_ty.toIntern(),
+                .base_addr = .{ .anon_decl = .{
+                    .val = array_val,
+                    .orig_ty = ptr_ty.toIntern(),
+                } },
+                .byte_offset = 0,
+            } });
+        },
+        .struct_init_one,
+        .struct_init_one_comma,
+        .struct_init_dot_two,
+        .struct_init_dot_two_comma,
+        .struct_init_dot,
+        .struct_init_dot_comma,
+        .struct_init,
+        .struct_init_comma,
+        => {
+            var buf: [2]Ast.Node.Index = undefined;
+            const struct_init = self.file.tree.fullStructInit(&buf, node).?;
+            if (struct_init.ast.type_expr != 0) {
+                return self.fail(.{ .node_abs = struct_init.ast.type_expr }, "type expressions not allowed in ZON", .{});
+            }
+            const types = try gpa.alloc(InternPool.Index, struct_init.ast.fields.len);
+            defer gpa.free(types);
```

Comment:
As before, these should probably go into the Sema arena. (Likewise with the other gpa usages here.)